rust/src/storage/s3/dynamodb_lock.rs

```diff
@@ -0,0 +1,510 @@
+    pub is_released: bool,
+    /// Optional data associated with this lock.
+    pub data: Option<String>,
+    /// The last time this lock was updated or retrieved.
+    pub lookup_time: u128,
+}
+
+impl LockItem {
+    fn is_expired(&self) -> bool {
+        if self.is_released {
+            return true;
+        }
+        now_millis() - self.lookup_time > (self.lease_duration as u128) * 1000
+    }
+}
+
+/// Error returned by the [`DynamoDbLockClient`] API.
+#[derive(thiserror::Error, Debug)]
+pub enum DynamoError {
+    /// Error caused by the DynamoDB table not being created.
+    #[error("Dynamo table not found")]
+    TableNotFound,
+
+    /// Error that indicates the condition in the DynamoDB operation could not be evaluated.
+    /// Mostly used by [`DynamoDbLockClient::acquire_lock`] to handle unsuccessful retries
+    /// of acquiring the lock.
+    #[error("Conditional check failed")]
+    ConditionalCheckFailed,
+
+    /// Error that is returned by [`DynamoDbLockClient::release_lock`] when the given lock
+    /// has been already expired and could not be released.
+    #[error("Lock is expired")]
+    LockIsExpired,
+
+    /// The required field of [`LockItem`] is missing in DynamoDB record or has incompatible type.
+    #[error("DynamoDB item has invalid schema")]
+    InvalidItemSchema,
+
+    /// Error returned by [`DynamoDbLockClient::acquire_lock`] which indicates that the lock could
+    /// not be acquired for more that returned number of seconds.
+    #[error("Could not acquire lock for {0} sec")]
+    TimedOut(u64),
+
+    /// Error caused by the [`DynamoDbClient::put_item`] request.
+    #[error("Put item error: {0}")]
+    PutItemError(RusotoError<PutItemError>),
+
+    /// Error caused by the [`DynamoDbClient::update_item`] request.
+    #[error("Update item error: {0}")]
+    UpdateItemError(RusotoError<UpdateItemError>),
+
+    /// Error caused by the [`DynamoDbClient::get_item`] request.
+    #[error("Get item error: {0}")]
+    GetItemError(RusotoError<GetItemError>),
+}
+
+impl From<RusotoError<PutItemError>> for DynamoError {
+    fn from(error: RusotoError<PutItemError>) -> Self {
+        match error {
+            RusotoError::Service(PutItemError::ConditionalCheckFailed(_)) => {
+                DynamoError::ConditionalCheckFailed
+            }
+            _ => DynamoError::PutItemError(error),
+        }
+    }
+}
+
+impl From<RusotoError<GetItemError>> for DynamoError {
+    fn from(error: RusotoError<GetItemError>) -> Self {
+        match error {
+            RusotoError::Service(GetItemError::ResourceNotFound(_)) => DynamoError::TableNotFound,
+            _ => DynamoError::GetItemError(error),
+        }
+    }
+}
+
+impl From<RusotoError<UpdateItemError>> for DynamoError {
+    fn from(error: RusotoError<UpdateItemError>) -> Self {
+        match error {
+            RusotoError::Service(UpdateItemError::ConditionalCheckFailed(_)) => {
+                DynamoError::LockIsExpired
```

Comment:
The conditional put could fail due to other constraint failures right? For example rvn or data field not matched?

Reply:
Okay yes, this one is questionable. Because reference implementation returns a boolean value where false catches all of the cases, basically when condition is failed. I'll change it that way.

Also this `From` is invalid because if we'd add other UpdateItem this'll be broken and misleading. I'll move this sort of pattern matching into `release_lock` function